Refactoring Task Infrastructure

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

This pull request makes a number of changes to our task orchestration infrastructure in order to introduce consistency and significantly improve cache performance. Generally speaking, from the user-perspective, there should be very little changed by this pull request.

**Note: I chose to batch the changes in this PR so that it will have a minimal impact on developer experience by being merged at the same time.**

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/40954, https://github.com/woocommerce/woocommerce/issues/40952, https://github.com/woocommerce/woocommerce/issues/40953.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Verify that CI is passing. All of the changes in this PR apply to CI as well.
2. `pnpm --filter='@woocommerce/plugin-woocommerce' build`
3. `pnpm --filter='@woocommerce/plugin-woocommerce' lint`
4. `pnpm --filter='@woocommerce/plugin-woocommerce' build`
5. Explore! Play around with different commands and see how things work out.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

ObliviousHarmony committed:
Renamed `woocommerce/client/admin` Package

NPM package names _must_ either be single-word or have
a scope. The current structure is unsupported and has been
changed. The new package name is `@woocommerce/admin-library`.

ObliviousHarmony committed:
Renamed `woocommerce/client/legacy` Package

NPM requires names to be single-word or have a scope. The
current name is not valid. The package has been renamed
`@woocommerce/shortcode-assets` to better match its purpose.

ObliviousHarmony committed:
Renamed WooCommerce Plugin Package

In order to prevent supply-chain attacks we have renamed
the package. The new name is `@woocommerce/plugin-woocommerce`.

ObliviousHarmony committed:
Renamed WooCommerce Beta Tester Package

As part of the prevention of supply chain attacks we are
renaming the package `@woocommerce/plugin-woocommerce-beta-tester`.

ObliviousHarmony committed:
Fixed WooCommerce Plugin Name Documentation

ObliviousHarmony committed:
Renamed WooCommerce AI Plugin Package

As part of preventing supply chain attacks, the package was
renamed `@woocommerce/plugin-woo-ai`.

ObliviousHarmony committed:
Renamed WooCommerce Docs Plugin Package

As part of the prevention of supply chain attacks the package has been
renamed to `@woocommerce/plugin-woocommerce-docs`.

ObliviousHarmony committed:
Removed Unnecessary PNPM Overrides

Since we are using `syncpack`, we have no reason to use the overrides.

ObliviousHarmony committed:
Lock File Update

ObliviousHarmony committed:
Removed Turborepo

ObliviousHarmony committed:
Removed Unnecessary `.gitignore` Files

ObliviousHarmony committed:
Added `wireit` Dependency

We're adding this to _every_ package since they have their own
configuration and scripts to run. This also keeps them
isolated so that they can operate independently.

ObliviousHarmony committed:
Privatized Plugin Packages

ObliviousHarmony committed:
Sorted `package.json` Files

ObliviousHarmony committed:
Added Missing Core Plugin Dependencies

ObliviousHarmony committed:
Added Wireit Dependency PNPM Hook

We can use a PNPM hook to make sure that the dependencies
are accounted for in the fingerprint. This script scans through
all of the packages with Wireit and adds the outputs for
linked dependencies.

ObliviousHarmony committed:
Added Wireit: @woocommerce/admin-e2e-tests

ObliviousHarmony committed:
Added Wireit: @woocommerce/admin-layout

ObliviousHarmony committed:
Changed `build:project:js` To `build:project:esm`

ObliviousHarmony committed:
Fixed Invalid TSConfig Files

ObliviousHarmony committed:
Reverted Wireit Package Changes

We're going to use an automatic script to make these changes.

ObliviousHarmony committed:
Sorted Package Files

ObliviousHarmony committed:
Add Wireit: Lots of packages!

ObliviousHarmony committed:
Add Wireit: Remaining Packages

ObliviousHarmony committed:
Add Wireit: `@woocommerce/admin-library`

ObliviousHarmony committed:
Fixed Test Commands

ObliviousHarmony committed:
Revised Test Command

We will never run all tests concurrently so this is better.

ObliviousHarmony committed:
Add Wireit: `@woocommerce/shortcode-assets`

ObliviousHarmony committed:
Added Wireit Commands to WooCommerce

There's quite a bit of work needed in this file but
I'll come back to that later when I can review it all
in more detail.

ObliviousHarmony committed:
Removed Turbo References

ObliviousHarmony committed:
Fixed Dependency Hook

ObliviousHarmony committed:
Wireit Dependencies Set

ObliviousHarmony committed:
Updated Monorepo Setup

ObliviousHarmony committed:
Updated Root Package

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Removed Unnecessary Caching

ObliviousHarmony committed:
Cache Fix Attempt

ObliviousHarmony committed:
Cache Fix Attempt

ObliviousHarmony committed:
Fixed Matrix Generation Failure Evaluation

ObliviousHarmony committed:
Cache Fix Attempt

ObliviousHarmony committed:
Fixed Test Command

ObliviousHarmony committed:
Applied Consistent Outputs

ObliviousHarmony committed:
Change Asset Build Outputs

Wireit doesn't allow us to have build outputs in a directory outside
of the package. We can get around this by having a build step in
WooCommerce that copies the files from the output directories.

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Removed Turbo Reference

ObliviousHarmony committed:
Refactored Monorepo Setup Action

This makes some adjustments for consistency in style as well as making
improvements to usability and performance. It also updates all of the
pinned hashes to pull in any fixes that have been made.

ObliviousHarmony committed:
Fixed Wrong Equality Operator

Too used to strict equality I guess :)

ObliviousHarmony committed:
More Fixes

ObliviousHarmony committed:
Added Required Shell

ObliviousHarmony committed:
Fix PNPM Install Caching

ObliviousHarmony committed:
Debug Logging

ObliviousHarmony committed:
Moved Input Parsing

It looks like it may have been too late and not getting parsed?

ObliviousHarmony committed:
Testing Fix

ObliviousHarmony committed:
Updated `setup-woocommerce-monorepo` Usage

ObliviousHarmony committed:
Updated Documentation

rrennick commented:
@ObliviousHarmony There's lots of merge conflicts with this. Is this something that you will need to resolve just prior to merging?

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony commented:
I fixed the conflicts @rrennick. The joy of such a large PR :smile:

rrennick commented:
I noticed that linting logged errors in the `build` folder for blocks:

```
FILE: build/woocommerce/packages/woocommerce-blocks/build/stock-filter.asset.php
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 1 LINE
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1 | ERROR | [ ] Missing file doc comment (Squiz.Commenting.FileComment.Missing)
 1 | ERROR | [x] Missing space after array opener. (WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceAfterArrayOpener)
 1 | ERROR | [x] Missing space before array closer. (WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceBeforeArrayCloser)
 1 | ERROR | [x] When a multi-item array uses associative keys, each value should start on a new line.
   |       |     (WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound)
```

These probably don't exist in the blocks repo but post merge these files will exist in a local dev environment and should be excluded from linting. 

As it is, should the lint be processing files in `plugins/woocommerce/packages`at all as the contents of the folder is pulled in from other repositories?

rrennick commented:
The pnpm commands in https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/performance/README.md have not been updated.

rrennick commented on packages/js/e2e-environment/external.md:
```diff
@@ -38,7 +38,7 @@ Again, if you don't have shell access to your test site, through WP Admin ensure
 4. You have a customer user set up named 'Jane Smith'. This user should be a `subscriber` and again make sure their username and password are reflected in `/plugins/woocommerce/tests/e2e/config/default.json`.
 
-You should then be able to run the e2e tests by running `pnpm run e2e --filter=woocommerce`.
+You should then be able to run the e2e tests by running `pnpm run e2e --filter=@woocommerce/plugin-woocommerce`.
```
Adding a note that I created #41649 to address the use of monorepo specific commands in these `.md` files.

rrennick commented on plugins/woocommerce/tests/performance/README.md:
````diff
@@ -57,8 +57,8 @@ Before using the tests a test environment is needed to run against.
 ```sh
-pnpm run env:dev --filter=woocommerce
-pnpm run env:performance-init --filter=woocommerce
+pnpm run env:dev --filter=@woocommerce/plugin-woocommerce
````
This command is loading e2e-environment instead of wp-env.

ObliviousHarmony replied:
That script runs `"env:dev": "pnpm wp-env start --update"` and should be setting up `wp-env`.

rrennick replied:
Also, curious about why 3 `plugins/woocommerce/tests/*/README.md` were deleted.

rrennick replied:
> That script runs "env:dev": "pnpm wp-env start --update" and should be setting up wp-env.

`run` needs to be removed.

```suggestion
pnpm env:dev --filter=@woocommerce/plugin-woocommerce
```

rrennick commented on plugins/woocommerce/tests/performance/README.md:
```diff
@@ -57,8 +57,8 @@ Before using the tests a test environment is needed to run against.
-pnpm run env:dev --filter=woocommerce
-pnpm run env:performance-init --filter=woocommerce
+pnpm run env:dev --filter=@woocommerce/plugin-woocommerce
+pnpm run env:performance-init --filter=@woocommerce/plugin-woocommerce
```
`run` should be removed here as well.

rrennick reviewed (commented):
@jonathansadowski I've left a few comments and created some issues that weren't directly related to this PR. I did not find any functional issues in my testing of build and testing scripts. I did not test any of the release related scripting.

@ObliviousHarmony Nice work. Other than the comments I left it would be great to have this rebased to current `trunk` and have CI pass before merging. IIRC, this branch is about 150 commits behind trunk.

jonathansadowski commented:
> @jonathansadowski I've left a few comments and created some issues that weren't directly related to this PR. I did not find any functional issues in my testing of build and testing scripts. I did not test any of the release related scripting.

I've tested some of the release-related things... namely building release zips, as well as the tooling used for code-freeze, none of that broke, though I wouldn't expect the latter to.

I also didn't discover any functional issues, though I still have a list of things I wanted to check, and I'm not sure I will ahead of the holiday weekend, but that list is entirely related to the changes to `.github/actions/setup-woocommerce-monorepo/action.yml`, mostly because I recall previous changes to this action causing some unforeseen issues.

I didn't notice any functional issues with anything I've checked so far, and the things I wanted to check in the changes to `setup-woocommerce-monorepo` all appeared to be non-problematic at first glance (things like switching from string literals to boolean with the input parameters), so I think I'd be comfortable with this PR if like Ron said this were rebased to current trunk and CI were passing.

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Lock file update

ObliviousHarmony committed:
Fixed Beta Tester Release Workflow

ObliviousHarmony committed:
Removed Unnecessary `run` Usage

Since `pnpm` defaults to `run` unless a built-in
command exists we don't really need to use `run`.

ObliviousHarmony committed:
Removed Unnecessary `run` From `pnpm`

ObliviousHarmony committed:
Lock File Update

This takes the lock file from `trunk` and runs an install on it. This
should remove any accidental dependency
changes caused by the PR.

ObliviousHarmony committed:
Lock File Update

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Sorted Package Files

ObliviousHarmony committed:
Synchronized Dependencies

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/task-infrastructure

ObliviousHarmony committed:
Persist Package Newline After Dependency Parse

ObliviousHarmony committed:
Lock File Update

ObliviousHarmony committed:
Disabled CI Install Filtering

There is currently a bug with `pnpm` that needs fixed first.

ObliviousHarmony committed:
Remove Unused Tools
